File: Snooper/src/test/java/com/prateekj/snooper/networksnooper/helper/DataCopyHelperTest.java
package com.prateekj.snooper.networksnooper.helper;

import android.content.res.Resources;
import androidx.annotation.NonNull;

import com.prateekj.snooper.R;
import com.prateekj.snooper.formatter.ResponseFormatter;
import com.prateekj.snooper.formatter.ResponseFormatterFactory;
import com.prateekj.snooper.networksnooper.model.HttpCallRecord;
import com.prateekj.snooper.networksnooper.model.HttpHeader;
import com.prateekj.snooper.networksnooper.model.HttpHeaderValue;

import org.junit.Before;
import org.junit.Test;

import java.util.ArrayList;
import java.util.Collections;

import static com.prateekj.snooper.utils.TestUtilities.getDate;
import static java.util.Arrays.asList;
import static org.hamcrest.CoreMatchers.is;
import static org.junit.Assert.assertThat;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

public class DataCopyHelperTest {

  private ResponseFormatterFactory formatterFactory;
  private HttpCallRecord httpCall;
  private ResponseFormatter responseFormatter;
  private DataCopyHelper dataCopyHelper;
  private Resources resources;

  @Before
  public void setUp() throws Exception {
    formatterFactory = mock(ResponseFormatterFactory.class);
    httpCall = mock(HttpCallRecord.class);
    responseFormatter = mock(ResponseFormatter.class);
    resources = mock(Resources.class);
    dataCopyHelper = new DataCopyHelper(httpCall, formatterFactory, resources);
    mockStringResources();
  }

  @Test
  public void shouldCopyResponseWithoutFormattingIfContentHeadersNullInResponseData() throws Exception {
    String responseBody = "response body";
    when(httpCall.getResponseHeader("Content-Type")).thenReturn(null);
    when(httpCall.getResponseBody()).thenReturn(responseBody);
    when(formatterFactory.getFor("application/json")).thenReturn(responseFormatter);

    String responseDataForCopy = dataCopyHelper.getResponseDataForCopy();

    assertThat(responseDataForCopy, is(responseBody));
    verify(responseFormatter, never()).format(responseBody);
  }

  @Test
  public void shouldCopyResponseWithoutFormattingIfContentHeaderValuesIsMissingInResponseData() throws Exception {
    String responseBody = "response body";
    HttpHeader httpHeader = new HttpHeader();
    httpHeader.setValues(new ArrayList<HttpHeaderValue>());
    when(httpCall.getResponseHeader("Content-Type")).thenReturn(httpHeader);
    when(httpCall.getResponseBody()).thenReturn(responseBody);
    when(formatterFactory.getFor("application/json")).thenReturn(responseFormatter);

    String responseDataForCopy = dataCopyHelper.getResponseDataForCopy();

    assertThat(responseDataForCopy, is(responseBody));
    verify(responseFormatter, never()).format(responseBody);
  }

  @Test
  public void shouldCopyResponseWithFormattingIfContentHeadersPresentInResponseData() throws Exception {
    String responseBody = "response body";
    String formattedResponseBody = "formatted response body";
    when(httpCall.getResponseHeader("Content-Type")).thenReturn(getJsonContentTypeHeader());
    when(httpCall.getResponseBody()).thenReturn(responseBody);
    when(formatterFactory.getFor("application/json")).thenReturn(responseFormatter);
    when(responseFormatter.format(responseBody)).thenReturn(formattedResponseBody);

    String responseDataForCopy = dataCopyHelper.getResponseDataForCopy();

    assertThat(responseDataForCopy, is(formattedResponseBody));
    verify(responseFormatter).format(responseBody);
  }

  @Test
  public void shouldCopyEmptyStringWhenResponseIsNotPresent() throws Exception {
    when(httpCall.getResponseBody()).thenReturn(null);

    String responseDataForCopy = dataCopyHelper.getResponseDataForCopy();

    assertThat(responseDataForCopy, is(""));
  }

  @Test
  public void shouldCopyRequestWithoutFormattingIfContentHeaderIsNullInRequestData() throws Exception {
    String requestBody = "response body";
    when(httpCall.getRequestHeader("Content-Type")).thenReturn(null);
    when(httpCall.getPayload()).thenReturn(requestBody);
    when(formatterFactory.getFor("application/json")).thenReturn(responseFormatter);

    String requestDataForCopy = dataCopyHelper.getRequestDataForCopy();

    assertThat(requestDataForCopy, is(requestBody));
    verify(responseFormatter, never()).format(requestBody);
  }

  @Test
  public void shouldCopyRequestWithoutFormattingIfContentHeaderValuesMissingInRequestData() throws Exception {
    String requestBody = "response body";
    HttpHeader httpHeader = new HttpHeader();
    httpHeader.setValues(new ArrayList<HttpHeaderValue>());
    when(httpCall.getRequestHeader("Content-Type")).thenReturn(httpHeader);
    when(httpCall.getPayload()).thenReturn(requestBody);
    when(formatterFactory.getFor("application/json")).thenReturn(responseFormatter);

    String requestDataForCopy = dataCopyHelper.getRequestDataForCopy();

    assertThat(requestDataForCopy, is(requestBody));
    verify(responseFormatter, never()).format(requestBody);
  }

  @Test
  public void shouldCopyRequestResponseHeadersPresent() throws Exception {
    HttpHeader httpHeader = getAcceptLanguageHttpHeader();

    when(httpCall.getRequestHeaders()).thenReturn(asList(httpHeader, getJsonContentTypeHeader()));
    when(httpCall.getRequestHeader("Content-Type")).thenReturn(getJsonContentTypeHeader());
    when(httpCall.getResponseHeaders()).thenReturn(asList(httpHeader, getHeader()));
    when(httpCall.getResponseHeader("Content-Type")).thenReturn(getJsonContentTypeHeader());

    String copiedHeaders = dataCopyHelper.getHeadersForCopy();

    assertThat(copiedHeaders, is("\nRequest Headers\naccept-language: en-US,en;q=0.8,hi;q=0.6\n" +
      "Content-Type: application/json\n\nResponse Headers\naccept-language: en-US,en;q=0.8,hi;q=0.6\n" +
      "Header: headerValue\n"));
  }

  @Test
  public void shouldCopyOnlyRequestHeadersPresentIfResponseHeadersMissing() throws Exception {
    HttpHeader httpHeader = getAcceptLanguageHttpHeader();

    when(httpCall.getRequestHeaders()).thenReturn(asList(httpHeader, getJsonContentTypeHeader()));
    when(httpCall.getRequestHeader("Content-Type")).thenReturn(getJsonContentTypeHeader());

    String copiedHeaders = dataCopyHelper.getHeadersForCopy();

    assertThat(copiedHeaders, is("\nRequest Headers\naccept-language: en-US,en;q=0.8,hi;q=0.6\n" +
      "Content-Type: application/json\n"));
  }

  @Test
  public void shouldCopyOnlyResponseHeadersPresentIfRequestHeadersMissing() throws Exception {
    HttpHeader httpHeader = getAcceptLanguageHttpHeader();

    when(httpCall.getResponseHeaders()).thenReturn(asList(httpHeader, getHeader()));
    when(httpCall.getResponseHeader("Content-Type")).thenReturn(getJsonContentTypeHeader());

    String copiedHeaders = dataCopyHelper.getHeadersForCopy();

    assertThat(copiedHeaders, is("\nResponse Headers\naccept-language: en-US,en;q=0.8,hi;q=0.6\n" +
      "Header: headerValue\n"));
  }

  @Test
  public void shouldCopyEmptyStringWhenRequestIsNotPresent() throws Exception {
    when(httpCall.getResponseBody()).thenReturn(null);

    String requestDataForCopy = dataCopyHelper.getRequestDataForCopy();

    assertThat(requestDataForCopy, is(""));
  }

  @Test
  public void shouldAskViewToCopyTheError() throws Exception {
    String error = "error";
    when(httpCall.getError()).thenReturn(error);

    String errorsForCopy = dataCopyHelper.getErrorsForCopy();

    assertThat(errorsForCopy, is(error));
  }

  @Test
  public void shouldShareRequestResponseData() throws Exception {
    String requestBody = "request body";
    String formatRequestBody = "format Request body";
    String responseBody = "response body";
    String formatResponseBody = "format Response body";

    HttpHeader httpHeader = getAcceptLanguageHttpHeader();

    when(httpCall.getRequestHeaders()).thenReturn(asList(httpHeader, getJsonContentTypeHeader()));
    when(httpCall.getRequestHeader("Content-Type")).thenReturn(getJsonContentTypeHeader());
    when(httpCall.getPayload()).thenReturn(requestBody);
    when(httpCall.getResponseHeaders()).thenReturn(asList(httpHeader, getHeader()));
    when(httpCall.getResponseHeader("Content-Type")).thenReturn(getJsonContentTypeHeader());
    when(httpCall.getResponseBody()).thenReturn(responseBody);
    when(httpCall.getDate()).thenReturn(getDate(2017, 4, 12, 1, 2, 3));
    when(formatterFactory.getFor("application/json")).thenReturn(responseFormatter);
    when(responseFormatter.format(requestBody)).thenReturn(formatRequestBody);
    when(responseFormatter.format(responseBody)).thenReturn(formatResponseBody);

    StringBuilder httpCallData = dataCopyHelper.getHttpCallData();

    assertThat(httpCallData.toString(), is("Request Body\nformat Request body\nRequest Headers\naccept-language: en-US," +
      "en;q=0.8,hi;q=0.6\nContent-Type: application/json\nResponse Body\nformat Response body\nResponse Headers\n" +
      "accept-language: en-US,en;q=0.8,hi;q=0.6\nHeader: headerValue\n"));
    verify(responseFormatter, times(1)).format(requestBody);
    verify(responseFormatter, times(1)).format(responseBody);
  }

  @Test
  public void shouldShareResponseDataOnlyIfRequestDataEmpty() throws Exception {
    String responseBody = "response body";
    String formatResponseBody = "format Response body";

    when(httpCall.getResponseHeader("Content-Type")).thenReturn(getJsonContentTypeHeader());
    when(httpCall.getResponseBody()).thenReturn(responseBody);
    when(httpCall.getRequestHeaders()).thenReturn(null);
    when(httpCall.getDate()).thenReturn(getDate(2017, 4, 12, 1, 2, 3));
    when(formatterFactory.getFor("application/json")).thenReturn(responseFormatter);
    when(responseFormatter.format(responseBody)).thenReturn(formatResponseBody);

    StringBuilder httpCallData = dataCopyHelper.getHttpCallData();

    assertThat(httpCallData.toString(), is("Response Body\nformat Response body"));
    verify(responseFormatter, times(1)).format(responseBody);
  }

  @Test
  public void shouldShareRequestDataOnlyIfResponseDataEmpty() throws Exception {
    String requestBody = "request body";
    String formatRequestBody = "format Request body";

    when(httpCall.getRequestHeader("Content-Type")).thenReturn(getJsonContentTypeHeader());
    when(httpCall.getPayload()).thenReturn(requestBody);
    when(httpCall.getResponseBody()).thenReturn(null);
    when(httpCall.getDate()).thenReturn(getDate(2017, 4, 12, 1, 2, 3));
    when(formatterFactory.getFor("application/json")).thenReturn(responseFormatter);
    when(responseFormatter.format(requestBody)).thenReturn(formatRequestBody);

    StringBuilder httpCallData = dataCopyHelper.getHttpCallData();

    assertThat(httpCallData.toString(), is("Request Body\nformat Request body"));
    verify(responseFormatter, times(1)).format(requestBody);
  }

  @NonNull
  private HttpHeader getJsonContentTypeHeader() {
    HttpHeaderValue headerValue = new HttpHeaderValue("application/json");
    HttpHeader httpHeader = new HttpHeader("Content-Type");
    httpHeader.setValues(Collections.singletonList(headerValue));
    return httpHeader;
  }

  @NonNull
  private HttpHeader getHeader() {
    HttpHeaderValue headerValue = new HttpHeaderValue("headerValue");
    HttpHeader httpHeader = new HttpHeader("Header");
    httpHeader.setValues(Collections.singletonList(headerValue));
    return httpHeader;
  }

  @NonNull
  private HttpHeader getAcceptLanguageHttpHeader() {
    HttpHeader httpHeader = new HttpHeader("accept-language");
    HttpHeaderValue value1 = new HttpHeaderValue("en-US,en");
    HttpHeaderValue value2 = new HttpHeaderValue("q=0.8,hi");
    HttpHeaderValue value3 = new HttpHeaderValue("q=0.6");
    httpHeader.setValues(asList(value1, value2, value3));
    return httpHeader;
  }

  private void mockStringResources() {
    when(resources.getString(R.string.request_body_heading)).thenReturn("Request Body");
    when(resources.getString(R.string.request_headers)).thenReturn("Request Headers");
    when(resources.getString(R.string.response_body_heading)).thenReturn("Response Body");
    when(resources.getString(R.string.response_headers)).thenReturn("Response Headers");
  }
}

File: Snooper/src/test/java/com/prateekj/snooper/networksnooper/helper/DataCopyHelperTest.kt
package com.prateekj.snooper.networksnooper.helper

import android.content.res.Resources
import com.prateekj.snooper.R
import com.prateekj.snooper.formatter.ResponseFormatter
import com.prateekj.snooper.formatter.ResponseFormatterFactory
import com.prateekj.snooper.networksnooper.model.HttpCallRecord
import com.prateekj.snooper.networksnooper.model.HttpHeader
import com.prateekj.snooper.networksnooper.model.HttpHeaderValue
import com.prateekj.snooper.utils.TestUtils.getDate
import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import org.hamcrest.CoreMatchers.`is`
import org.junit.Assert.assertThat
import org.junit.Before
import org.junit.Test
import java.util.ArrayList

class DataCopyHelperTest {

  private lateinit var formatterFactory: ResponseFormatterFactory
  private lateinit var httpCall: HttpCallRecord
  private lateinit var responseFormatter: ResponseFormatter
  private lateinit var dataCopyHelper: DataCopyHelper
  private lateinit var resources: Resources

  private val jsonContentTypeHeader: HttpHeader
    get() {
      val headerValue = HttpHeaderValue("application/json")
      val httpHeader = HttpHeader("Content-Type")
      httpHeader.values = listOf(headerValue)
      return httpHeader
    }

  private val header: HttpHeader
    get() {
      val headerValue = HttpHeaderValue("headerValue")
      val httpHeader = HttpHeader("Header")
      httpHeader.values = listOf(headerValue)
      return httpHeader
    }

  private val acceptLanguageHttpHeader: HttpHeader
    get() {
      val httpHeader = HttpHeader("accept-language")
      val value1 = HttpHeaderValue("en-US,en")
      val value2 = HttpHeaderValue("q=0.8,hi")
      val value3 = HttpHeaderValue("q=0.6")
      httpHeader.values = listOf(value1, value2, value3)
      return httpHeader
    }

  @Before
  @Throws(Exception::class)
  fun setUp() {
    formatterFactory = mockk(relaxed = true)
    httpCall = mockk(relaxed = true)
    responseFormatter = mockk(relaxed = true)
    resources = mockk(relaxed = true)
    dataCopyHelper = DataCopyHelper(httpCall, formatterFactory, resources)
    mockStringResources()
  }

  @Test
  @Throws(Exception::class)
  fun shouldCopyResponseWithoutFormattingIfContentHeadersNullInResponseData() {
    val responseBody = "response body"
    every { httpCall.getResponseHeader("Content-Type") } returns null
    every { httpCall.responseBody } returns responseBody
    every { formatterFactory.getFor("application/json") } returns responseFormatter

    val responseDataForCopy = dataCopyHelper.getResponseDataForCopy()

    assertThat(responseDataForCopy, `is`(responseBody))
    verify(exactly = 0) { responseFormatter.format(responseBody) }
  }

  @Test
  @Throws(Exception::class)
  fun shouldCopyResponseWithoutFormattingIfContentHeaderValuesIsMissingInResponseData() {
    val responseBody = "response body"
    val httpHeader = HttpHeader()
    httpHeader.values = ArrayList()
    every { httpCall.getResponseHeader("Content-Type") } returns httpHeader
    every { httpCall.responseBody } returns responseBody
    every { formatterFactory.getFor("application/json") } returns responseFormatter

    val responseDataForCopy = dataCopyHelper.getResponseDataForCopy()

    assertThat(responseDataForCopy, `is`(responseBody))
    verify(exactly = 0) { responseFormatter.format(responseBody) }
  }

  @Test
  @Throws(Exception::class)
  fun shouldCopyResponseWithFormattingIfContentHeadersPresentInResponseData() {
    val responseBody = "response body"
    val formattedResponseBody = "formatted response body"
    every { httpCall.getResponseHeader("Content-Type") } returns jsonContentTypeHeader
    every { httpCall.responseBody } returns responseBody
    every { formatterFactory.getFor("application/json") } returns responseFormatter
    every { responseFormatter.format(responseBody) } returns formattedResponseBody

    val responseDataForCopy = dataCopyHelper.getResponseDataForCopy()

    assertThat(responseDataForCopy, `is`(formattedResponseBody))
    verify { responseFormatter.format(responseBody) }
  }

  @Test
  @Throws(Exception::class)
  fun shouldCopyEmptyStringWhenResponseIsNotPresent() {
    every { httpCall.responseBody } returns null

    val responseDataForCopy = dataCopyHelper.getResponseDataForCopy()

    assertThat(responseDataForCopy, `is`(""))
  }

  @Test
  @Throws(Exception::class)
  fun shouldCopyRequestWithoutFormattingIfContentHeaderIsNullInRequestData() {
    val requestBody = "response body"
    every { httpCall.getRequestHeader("Content-Type") } returns null
    every { httpCall.payload } returns requestBody
    every { formatterFactory.getFor("application/json") } returns responseFormatter

    val requestDataForCopy = dataCopyHelper.getRequestDataForCopy()

    assertThat(requestDataForCopy, `is`(requestBody))
    verify(exactly = 0) { responseFormatter.format(requestBody) }
  }

  @Test
  @Throws(Exception::class)
  fun shouldCopyRequestWithoutFormattingIfContentHeaderValuesMissingInRequestData() {
    val requestBody = "response body"
    val httpHeader = HttpHeader()
    httpHeader.values = ArrayList()
    every { httpCall.getRequestHeader("Content-Type") } returns httpHeader
    every { httpCall.payload } returns requestBody
    every { formatterFactory.getFor("application/json") } returns responseFormatter

    val requestDataForCopy = dataCopyHelper.getRequestDataForCopy()

    assertThat(requestDataForCopy, `is`(requestBody))
    verify(exactly = 0) { responseFormatter.format(requestBody) }
  }

  @Test
  @Throws(Exception::class)
  fun shouldCopyRequestResponseHeadersPresent() {
    val httpHeader = acceptLanguageHttpHeader

    every { httpCall.requestHeaders } returns listOf(httpHeader, jsonContentTypeHeader)
    every { httpCall.getRequestHeader("Content-Type") } returns jsonContentTypeHeader
    every { httpCall.responseHeaders } returns listOf(httpHeader, header)
    every { httpCall.getResponseHeader("Content-Type") } returns jsonContentTypeHeader

    val copiedHeaders = dataCopyHelper.getHeadersForCopy()

    assertThat(
      copiedHeaders, `is`(
        "\nRequest Headers\naccept-language: en-US,en;q=0.8,hi;q=0.6\n" +
          "Content-Type: application/json\n\nResponse Headers\naccept-language: en-US,en;q=0.8,hi;q=0.6\n" +
          "Header: headerValue\n"
      )
    )
  }

  @Test
  @Throws(Exception::class)
  fun shouldCopyOnlyRequestHeadersPresentIfResponseHeadersMissing() {
    val httpHeader = acceptLanguageHttpHeader

    every { httpCall.requestHeaders } returns listOf(httpHeader, jsonContentTypeHeader)
    every { httpCall.responseHeaders } returns listOf()
    every { httpCall.getRequestHeader("Content-Type") } returns jsonContentTypeHeader

    val copiedHeaders = dataCopyHelper.getHeadersForCopy()

    assertThat(
      copiedHeaders,
      `is`("\nRequest Headers\naccept-language: en-US,en;q=0.8,hi;q=0.6\n" + "Content-Type: application/json\n")
    )
  }

  @Test
  @Throws(Exception::class)
  fun shouldCopyOnlyResponseHeadersPresentIfRequestHeadersMissing() {
    val httpHeader = acceptLanguageHttpHeader

    every { httpCall.requestHeaders } returns listOf()
    every { httpCall.responseHeaders } returns listOf(httpHeader, header)
    every { httpCall.getResponseHeader("Content-Type") } returns jsonContentTypeHeader

    val copiedHeaders = dataCopyHelper.getHeadersForCopy()

    assertThat(
      copiedHeaders,
      `is`("\nResponse Headers\naccept-language: en-US,en;q=0.8,hi;q=0.6\n" + "Header: headerValue\n")
    )
  }

  @Test
  @Throws(Exception::class)
  fun shouldCopyEmptyStringWhenRequestIsNotPresent() {
    every { httpCall.responseBody } returns null

    val requestDataForCopy = dataCopyHelper.getRequestDataForCopy()

    assertThat(requestDataForCopy, `is`(""))
  }

  @Test
  @Throws(Exception::class)
  fun shouldAskViewToCopyTheError() {
    val error = "error"
    every { httpCall.error } returns error

    val errorsForCopy = dataCopyHelper.getErrorsForCopy()

    assertThat(errorsForCopy, `is`(error))
  }

  @Test
  @Throws(Exception::class)
  fun shouldShareRequestResponseData() {
    val requestBody = "request body"
    val formatRequestBody = "format Request body"
    val responseBody = "response body"
    val formatResponseBody = "format Response body"

    val httpHeader = acceptLanguageHttpHeader

    every { httpCall.requestHeaders } returns listOf(httpHeader, jsonContentTypeHeader)
    every { httpCall.getRequestHeader("Content-Type") } returns jsonContentTypeHeader
    every { httpCall.payload } returns requestBody
    every { httpCall.responseHeaders } returns listOf(httpHeader, header)
    every { httpCall.getResponseHeader("Content-Type") } returns jsonContentTypeHeader
    every { httpCall.responseBody } returns responseBody
    every { httpCall.date } returns getDate(2017, 4, 12, 1, 2, 3)
    every { formatterFactory.getFor("application/json") } returns responseFormatter
    every { responseFormatter.format(requestBody) } returns formatRequestBody
    every { responseFormatter.format(responseBody) } returns formatResponseBody

    val httpCallData = dataCopyHelper.getHttpCallData()

    assertThat(
      httpCallData.toString(), `is`(
        "Request Body\nformat Request body\nRequest Headers\naccept-language: en-US," +
          "en;q=0.8,hi;q=0.6\nContent-Type: application/json\nResponse Body\nformat Response body\nResponse Headers\n" +
          "accept-language: en-US,en;q=0.8,hi;q=0.6\nHeader: headerValue\n"
      )
    )
    verify(exactly = 1) { responseFormatter.format(requestBody)}
    verify(exactly = 1) { responseFormatter.format(responseBody)}
  }

  @Test
  @Throws(Exception::class)
  fun shouldShareResponseDataOnlyIfRequestDataEmpty() {
    val responseBody = "response body"
    val formatResponseBody = "format Response body"

    every { httpCall.getResponseHeader("Content-Type") } returns jsonContentTypeHeader
    every { httpCall.responseBody } returns responseBody
    every { httpCall.requestHeaders } returns null
    every { httpCall.responseHeaders } returns null
    every { httpCall.date } returns getDate(2017, 4, 12, 1, 2, 3)
    every { formatterFactory.getFor("application/json") } returns responseFormatter
    every { responseFormatter.format(responseBody) } returns formatResponseBody

    val httpCallData = dataCopyHelper.getHttpCallData()

    assertThat(httpCallData.toString(), `is`("Response Body\nformat Response body"))
    verify(exactly = 1) { responseFormatter.format(responseBody) }
  }

  @Test
  @Throws(Exception::class)
  fun shouldShareRequestDataOnlyIfResponseDataEmpty() {
    val requestBody = "request body"
    val formatRequestBody = "format Request body"

    every { httpCall.getRequestHeader("Content-Type") } returns jsonContentTypeHeader
    every { httpCall.payload } returns requestBody
    every { httpCall.responseBody } returns null
    every { httpCall.requestHeaders } returns null
    every { httpCall.responseHeaders } returns null
    every { httpCall.date } returns getDate(2017, 4, 12, 1, 2, 3)
    every { formatterFactory.getFor("application/json") } returns responseFormatter
    every { responseFormatter.format(requestBody) } returns formatRequestBody

    val httpCallData = dataCopyHelper.getHttpCallData()

    assertThat(httpCallData.toString(), `is`("Request Body\nformat Request body"))
    verify(exactly = 1) { responseFormatter.format(requestBody) }
  }

  private fun mockStringResources() {
    every { resources.getString(R.string.request_body_heading) } returns "Request Body"
    every { resources.getString(R.string.request_headers) } returns "Request Headers"
    every { resources.getString(R.string.response_body_heading) } returns "Response Body"
    every { resources.getString(R.string.response_headers) } returns "Response Headers"
  }
}